Выполнить задание 6го спринта.

fcevt opened the pull request:
Сделать историю просмотров неограниченной ро размеру. Избавиться от повторных просмотров.
Реализовать в InMemoryHistoryManager двусвязный список. Реализовать метод removeNode() работающий за О(1). Реализовать класс Node для связного списка из задач. Реализовать метод getTasks для перекладывания истории просмотра из связного списка в ArrayList. Реализовать метод linkLast для добавления просмотренных задач в конец связного списка. Добавить тесты для новой истории просмотров.

fcevt committed:
Выполнить задание 6го спринта.
Сделать историю просмотров неограниченной ро размеру.
Избавиться от повторных просмотров.
Реализовать в InMemoryHistoryManager двусвязный список.
Реализовать метод removeNode() работающий за О(1).
Реализовать класс Node для связного списка из задач.
Реализовать метод getTasks для перекладывания истории просмотра из связного списка в ArrayList.
Реализовать метод linkLast для добавления просмотренных задач в конец связного списка.
Добавить тесты для новой истории просмотров.

fcevt committed:
Выполнить задание 6го спринта.
Сделать историю просмотров неограниченной ро размеру.
Избавиться от повторных просмотров.
Реализовать в InMemoryHistoryManager двусвязный список.
Реализовать метод removeNode() работающий за О(1).
Реализовать класс Node для связного списка из задач.
Реализовать метод getTasks для перекладывания истории просмотра из связного списка в ArrayList.
Реализовать метод linkLast для добавления просмотренных задач в конец связного списка.
Добавить тесты для новой истории просмотров.

fcevt committed:
Выполнить задание 6го спринта.
Сделать историю просмотров неограниченной ро размеру.
Избавиться от повторных просмотров.
Реализовать в InMemoryHistoryManager двусвязный список.
Реализовать метод removeNode() работающий за О(1).
Реализовать класс Node для связного списка из задач.
Реализовать метод getTasks для перекладывания истории просмотра из связного списка в ArrayList.
Реализовать метод linkLast для добавления просмотренных задач в конец связного списка.
Добавить тесты для новой истории просмотров.

pattyku commented on src/com/yandex/tracker/servise/InMemoryHistoryManager.java:
```diff
@@ -1,29 +1,103 @@
-    private final List<Task> browsingHistory;
+    private Node head;
+    private Node tail;
+    public int size;
```
size уже не нужен и логика связанная с ним. Удаляем

pattyku commented on src/com/yandex/tracker/servise/InMemoryHistoryManager.java:
```diff
@@ -1,29 +1,103 @@
+    private Node head;
+    private Node tail;
+    public int size;
+    private Map<Integer, Node> mapForDelete;
```
добавить еще final

pattyku commented on src/com/yandex/tracker/servise/InMemoryHistoryManager.java:
```diff
@@ -1,29 +1,103 @@
     public InMemoryHistoryManager() {
-        browsingHistory = new LinkedList<>();
+        head = null;
+        tail = null;
```
head и tail равны уже null, поэтому дополнительно присваивать значение не нужно

pattyku commented on src/com/yandex/tracker/servise/InMemoryHistoryManager.java:
```diff
@@ -1,29 +1,103 @@
+
+    //добавление просмотренных тасков в конец истории просмотров
+    Node linkLast(Task task) {
+        if (tail == null) {
```
здесь можно на много проще все написать
Во первых вынести создание ноды в начало. Если tail равен null, по умолчанию сохраниться null

`final Node node = new Node(task, tail, null);`

А дальше проверить равен ли head null

```
		if (head == null) {
			head = node;
		} else {
			tail.next = node;
		}
		tail = node;
```

pattyku commented on src/com/yandex/tracker/servise/HistoryManager.java:
```diff
@@ -4,9 +4,12 @@
+
+    List<Task> getHistory();
+
+    void removeNode(int id);
```
метод remove, в классе InMemoryHistoryManager дополнительно должен быть создан private метод removeNode, который вызывается в этом методе

pattyku commented on src/com/yandex/tracker/servise/InMemoryHistoryManager.java:
```diff
@@ -1,29 +1,103 @@
+    //удаление ноды в зависимости от ее положения вв списке
+    @Override
+    public void removeNode(int id) {
+        if (!mapForDelete.containsKey(id)) {
```
Эту проверку лучше заменить на 


```
if (node == null) {
	return;
}
```

pattyku commented on src/com/yandex/tracker/servise/InMemoryHistoryManager.java:
```diff
@@ -1,29 +1,103 @@
+
+    //удаление ноды в зависимости от ее положения вв списке
+    @Override
+    public void removeNode(int id) {
```
Делаем этот метод private и на вход в этот метод передаем ноду. Если придерживаться принципа разделения ответственности и то что это должен быть универсальный метод, он ничего не должен знать о mapForDelete. Вместо mapForDelete может быть добавлено допустим другое хранилище и этот метод должен уметь работать и с ним
`public void removeNode(Node node) {}`

И также нужно реализовать метод remove из интерфейса

```
@Override
public void remove(int id) {
    ....
}
```

pattyku commented on src/com/yandex/tracker/servise/InMemoryHistoryManager.java:
```diff
@@ -1,29 +1,103 @@
+        }
+        if (mapForDelete.containsKey(task.getId())) {
+            removeNode(task.getId());
+            mapForDelete.put(task.getId(), linkLast(task));
```
Эта строка выполняется независимо от условия, просто ее дублируешь. Ее нужно вынести за условие и удалить блок else. Внутри метода removeNode, поэтому здесь и дополнительная проверка не нужна


```
if (task == null) {
	 return;
}
 final int id = task.getId();
 removeNode(mapForDelete.remove(id));
 mapForDelete.put(id, linkLast(task));
```

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -54,12 +55,18 @@ public void removeTasks() {
 
```
В методе removeTasks() тоже нужно удалить задачи из истории

pattyku commented on src/com/yandex/tracker/servise/InMemoryTaskManager.java:
```diff
@@ -54,12 +55,18 @@ public void removeTasks() {
+        for (Integer id : epics.keySet()) {
+            browsingHistory.removeNode(id);
+        }
+        removeSubtasks();
```
Не используем одни публичные методы в других. Методы вынесенные в интерфейс это своего рода АПИ (пройдете дальше), где не допустим такой подход
В методе removeSubtasks() выполняется дополнительная логика с эпиком, которую здесь не нужно делать.
Нужно как и эпики пройтись циклом и удалить

fcevt committed:
В классе InMemoryHistoryManager:
Удалить поле size и всю логику связанную с ним.
Сделать final поле mapForDelete.
Отменить присваивание null полям head и tail при создании менеджера.
Упростить логику метода linkLast.
Создать метод remove(id) и private метод removeNode(Node) и реализовать их.
Изменить метод add.
В классе InMemoryTaskManager:
Дополнить метод removeTasks.
Изменить метод removeSubtasks.

fcevt committed:
В классе InMemoryHistoryManager:
Удалить поле size и всю логику связанную с ним.
Сделать final поле mapForDelete.
Отменить присваивание null полям head и tail при создании менеджера.
Упростить логику метода linkLast.
Создать метод remove(id) и private метод removeNode(Node) и реализовать их.
Изменить метод add.
В классе InMemoryTaskManager:
Дополнить метод removeTasks.
Изменить метод removeSubtasks.

pattyku commented on src/com/yandex/tracker/servise/InMemoryHistoryManager.java:
```diff
@@ -1,30 +1,89 @@
-                browsingHistory.removeFirst();
-            }
-            browsingHistory.add(task);
+    public void remove(int id) {
```
уже есть проверка в removeNode() дополнительно здесь можно не проверять. Можно упростить в одну строку заменив get на remove

```
public void remove(int id) {
      removeNode(mapForDelete.remove(id));
}
```
